vignettes/wrappers.Rmd

```diff
@@ -0,0 +1,151 @@
+---
+title: "How to write a function that wraps glue"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{How to write a function that wraps glue}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+library(glue)
+```
+
+Imagine that you want to call `glue()` repeatedly inside your own code (e.g. in your own package) with a non-default value for one or more arguments.
+For example, maybe you anticipate producing R code where `{` and `}` have specific syntactic meaning.
+Therefore, you'd prefer to use `<<` and `>>` as the opening and closing delimiters for expressions in `glue()`.
+
+Spoiler alert: here's the correct way to write such a wrapper:
+
+```{r}
+myglue <- function(..., .envir = parent.frame()) {
+  glue(..., .open = "<<", .close = ">>", .envir = .envir)
+}
+```
+
+This is the key move:
+
+> Include `.envir = parent.frame()` as an argument of the wrapper function and pass this `.envir` to the `.envir` argument of `glue()`.
+
+If you'd like to understand why this is the way, keep reading.
+
+## A first attempt that does not work
+
+Here's a simple attempt at writing a wrapper around `glue()`:
+
+```{r}
+myglue0 <- function(...) {
+  glue(..., .open = "@@", .close = "~~")
+}
+```
+
+From superficial experimentation, `myglue0()` appears to work:
+
+```{r}
+fn_def <- "
+  @@NAME~~ <- function(x) {
+    @@BODY~~
+  }"
+
+myglue0(fn_def, NAME = "one_plus_one", BODY = "1 + 1")
+```
+
+However slightly more sophisticated and realistic usage of `myglue0()` reveals a big problem.
+Here we use `myglue0()` inside another function, `fn_builder0()`:
+
+```{r, error = TRUE}
+fn_builder0 <- function(NAME, BODY) {
+  fn_def <- "
+    @@NAME~~ <- function(x) {
+      @@BODY~~
+    }"
+  myglue0(fn_def)
+}
+
+fn_builder0("two_times_two", "2 * 2")
+```
+
+What do you mean `NAME` is not found?!?
+It's one of the arguments of `fn_builder0()`.
+It should be "two_times_two".
+
+## Where does `glue()` evaluate code?
+
+What's going on?
+It's time to look at the (redacted) signature of `glue()`:
+
+```{r, eval = FALSE}
+glue(..., .envir = parent.frame(), ...)
+```
+
+The expressions inside a glue string are evaluated with respect to `.envir`, which defaults to the environment where `glue()` is called from.
+
+Let's make new artificial versions of our functions that make it easy to tell where the inner `glue()` call is getting its values.
+
+```{r}
+myglue1 <- function(...) {
```

Comment:
I think this would be easier to explain first, before you introduce all the complexities of a real example. Maybe something like this:

```R
x <- 0

myglue1 <- function(...) {
  x <- 1
  myglue2(...)
}

myglue2 <- function(...) {
  x <- 2
  glue::glue(...)
}

myglue2("{x}")
```

Reply:
Or maybe:

```R
x <- 0
y <- 0
z <- 0

myglue1 <- function(...) {
  x <- 1
  y <- 1
  myglue2(...)
}

myglue2 <- function(...) {
  x <- 2
  glue::glue(...)
}

myglue2("{x} {y} {z}")
```